<h1> Pandas
    
creating and reading and writing

In [1]:
import pandas as pd

There are two core objects in pandas: the DataFrame and the Series.
    
A DataFrame is a table. It contains an array of individual entries, each of which has a certain value. Each entry corresponds to a row (or record) and a column.

A Series, by contrast, is a sequence of data values. If a DataFrame is a table, a Series is a list. And in fact you can create one with nothing more than a list:

In [3]:
df = pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 
              'Sue': ['Pretty good.', 'Bland.']},
             index=['Product A', 'Product B'])
df

Bob           Sue
Product A    I liked it.  Pretty good.
Product B  It was awful.        Bland.

In [7]:
df2 = pd.Series([1,2,3,4])
df2

0    1
1    2
2    3
3    4
dtype: int64

A Series is, in essence, a single column of a DataFrame. So you can assign row labels to the Series the same way as before, using an index parameter. However, a Series does not have a column name, it only has one overall name:

In [4]:
pd.Series([30, 35, 40], index=['2015 Sales', '2016 Sales', '2017 Sales'], name='Product A')

2015 Sales    30
2016 Sales    35
2017 Sales    40
Name: Product A, dtype: int64

# Indexing and Selecting
# iloc vs loc

Index-based selection

Pandas indexing works in one of two paradigms. The first is index-based selection: selecting data based on its numerical position in the data. iloc follows this paradigm.

To select the first row of data in a DataFrame, we may use the following:

In [6]:
import pandas as pd
df = pd.read_csv("employees.csv")
df

First Name  Gender  Start Date Last Login Time  Salary  Bonus %  \
0      Douglas    Male    8/6/1993        12:42 PM   97308    6.945   
1       Thomas    Male   3/31/1996         6:53 AM   61933    4.170   
2        Maria  Female   4/23/1993        11:17 AM  130590   11.858   
3        Jerry    Male    3/4/2005         1:00 PM  138705    9.340   
4        Larry    Male   1/24/1998         4:47 PM  101004    1.389   
..         ...     ...         ...             ...     ...      ...   
995      Henry     NaN  11/23/2014         6:09 AM  132483   16.655   
996    Phillip    Male   1/31/1984         6:30 AM   42392   19.675   
997    Russell    Male   5/20/2013        12:39 PM   96914    1.421   
998      Larry    Male   4/20/2013         4:45 PM   60500   11.985   
999     Albert    Male   5/15/2012         6:24 PM  129949   10.169   

    Senior Management                  Team  
0                True             Marketing  
1                True                   NaN  
2               False               Finance  
3                True               Finance  
4                True       Client Services  
..                ...                   ...  
995             False          Distribution  
996             False               Finance  
997             False               Product  
998             False  Business Development  
999              True                 Sales  

[1000 rows x 8 columns]

In [7]:
df.iloc[:,0] #ทุกแถวที่อยู่ใน column 0

0      Douglas
1       Thomas
2        Maria
3        Jerry
4        Larry
        ...   
995      Henry
996    Phillip
997    Russell
998      Larry
999     Albert
Name: First Name, Length: 1000, dtype: object

Both loc and iloc are row-first, column-second. This is the opposite of what we do in native Python, which is column-first, row-second.

This means that it's marginally easier to retrieve rows, and marginally harder to get retrieve columns. To get a column with iloc, we can do the following:

` df.iloc[row, column]`

In [8]:
df.iloc[0,:]['First Name']#ทุกcolumn ที่อยู่ในแถว0 และเลือกเฉพาะ column ที่มีขื่อว่า First Name

'Douglas'

In [9]:
df.iloc[1,2]#แถว1หลัก2

'3/31/1996'

In [10]:
df.iloc[[14,2,3],:] #เช้าถึงแบบlist

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
14   Kimberly  Female  1/14/1999         7:13 AM   41426   14.543   
2       Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3       Jerry    Male   3/4/2005         1:00 PM  138705    9.340   

   Senior Management     Team  
14              True  Finance  
2              False  Finance  
3               True  Finance

In [24]:
ds = df.iloc[[14,2,3],:] #เข้าถึงแบบlist
ds.index = ["First", "second","third"]#เปลี่นยนชื่อ index
ds

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
First    Kimberly  Female  1/14/1999         7:13 AM   41426   14.543   
second      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
third       Jerry    Male   3/4/2005         1:00 PM  138705    9.340   

       Senior Management     Team  
First               True  Finance  
second             False  Finance  
third               True  Finance

Label-based selection

The second paradigm for attribute selection is the one followed by the loc operator: label-based selection. In this paradigm, it's the data index value, not its position, which matters.

For example, to get the first entry in reviews, we would now do the following:

In [25]:
df

First Name  Gender  Start Date Last Login Time  Salary  Bonus %  \
0      Douglas    Male    8/6/1993        12:42 PM   97308    6.945   
1       Thomas    Male   3/31/1996         6:53 AM   61933    4.170   
2        Maria  Female   4/23/1993        11:17 AM  130590   11.858   
3        Jerry    Male    3/4/2005         1:00 PM  138705    9.340   
4        Larry    Male   1/24/1998         4:47 PM  101004    1.389   
..         ...     ...         ...             ...     ...      ...   
995      Henry     NaN  11/23/2014         6:09 AM  132483   16.655   
996    Phillip    Male   1/31/1984         6:30 AM   42392   19.675   
997    Russell    Male   5/20/2013        12:39 PM   96914    1.421   
998      Larry    Male   4/20/2013         4:45 PM   60500   11.985   
999     Albert    Male   5/15/2012         6:24 PM  129949   10.169   

    Senior Management                  Team  
0                True             Marketing  
1                True                   NaN  
2               False               Finance  
3                True               Finance  
4                True       Client Services  
..                ...                   ...  
995             False          Distribution  
996             False               Finance  
997             False               Product  
998             False  Business Development  
999              True                 Sales  

[1000 rows x 8 columns]

In [11]:
df.loc[0:10,['Gender','Team']]

Gender                  Team
0     Male             Marketing
1     Male                   NaN
2   Female               Finance
3     Male               Finance
4     Male       Client Services
5     Male                 Legal
6   Female               Product
7   Female               Finance
8   Female           Engineering
9   Female  Business Development
10  Female                   NaN

Choosing between loc and iloc

When choosing or transitioning between loc and iloc, there is one "gotcha" worth keeping in mind, which is that the two methods use slightly different indexing schemes.

iloc uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So 0:10 will select entries 0,...,9. loc, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.

Why the change? Remember that loc can index any stdlib type: strings, for example. If we have a DataFrame with index values Apples, ..., Potatoes, ..., and we want to select "all the alphabetical fruit choices between Apples and Potatoes", then it's a lot more convenient to index df.loc['Apples':'Potatoes'] than it is to index something like df.loc['Apples', 'Potatoet'] (t coming after s in the alphabet).

This is particularly confusing when the DataFrame index is a simple numerical list, e.g. 0,...,1000. In this case df.iloc[0:1000] will return 1000 entries, while df.loc[0:1000] return 1001 of them! To get 1000 elements using loc, you will need to go one lower and ask for df.loc[0:999].

Otherwise, the semantics of using loc are the same as those for iloc.

เอาง่ายๆ iloc เข้าถึงค่าแบบ list indicing แบบรูปนี้

<img src="https://media.geeksforgeeks.org/wp-content/uploads/List-Slicing.jpg" alt="W3Schools.com">

In [12]:
df.loc[0:900] #ชื่อแถวที่ชื่อว่า 0 ถึง ชื่อแถวที่ชื่อว่า 900

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0      Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
1       Thomas    Male  3/31/1996         6:53 AM   61933    4.170   
2        Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3        Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4        Larry    Male  1/24/1998         4:47 PM  101004    1.389   
..         ...     ...        ...             ...     ...      ...   
896     Victor    Male  3/10/2010         5:01 PM   45267    3.942   
897    Kenneth     NaN  2/28/1994        10:10 AM   95296   10.146   
898       Jane  Female  4/27/2012         2:12 PM   59680   15.211   
899     Walter    Male  5/21/1992        12:39 AM  144701   16.323   
900  Christina  Female  6/23/2002         3:18 PM   35477   18.178   

    Senior Management                  Team  
0                True             Marketing  
1                True                   NaN  
2               False               Finance  
3                True               Finance  
4                True       Client Services  
..                ...                   ...  
896              True                 Sales  
897             False               Finance  
898              True  Business Development  
899              True             Marketing  
900             False       Human Resources  

[901 rows x 8 columns]

In [13]:
df.iloc[0:900] #index ที่ 0ถึง index ที่ 900

First Name  Gender  Start Date Last Login Time  Salary  Bonus %  \
0      Douglas    Male    8/6/1993        12:42 PM   97308    6.945   
1       Thomas    Male   3/31/1996         6:53 AM   61933    4.170   
2        Maria  Female   4/23/1993        11:17 AM  130590   11.858   
3        Jerry    Male    3/4/2005         1:00 PM  138705    9.340   
4        Larry    Male   1/24/1998         4:47 PM  101004    1.389   
..         ...     ...         ...             ...     ...      ...   
895     Janice     NaN  11/19/1991         6:02 PM  139791   16.968   
896     Victor    Male   3/10/2010         5:01 PM   45267    3.942   
897    Kenneth     NaN   2/28/1994        10:10 AM   95296   10.146   
898       Jane  Female   4/27/2012         2:12 PM   59680   15.211   
899     Walter    Male   5/21/1992        12:39 AM  144701   16.323   

    Senior Management                  Team  
0                True             Marketing  
1                True                   NaN  
2               False               Finance  
3                True               Finance  
4                True       Client Services  
..                ...                   ...  
895             False  Business Development  
896              True                 Sales  
897             False               Finance  
898              True  Business Development  
899              True             Marketing  

[900 rows x 8 columns]

In [14]:
x = [0,1,2,3,4,5,6,7,8,9,10]
x[0:10]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

สรุป

loc : การที่จะใช้ loc นั้น จะต้องระบุชื่อของแถวและ ชื่อของคอลัมน์ลงไปตรงๆ

dataframe.loc[‘index_name’ , ‘column_name’]

iloc : การที่จะใช้ iloc นั้นไม่สนใจชื่อ แต่สนใจตำแหน่งของแถว และ ตำแหน่งของคอลัมน์

dataframe.iloc[‘row_index_number’ , ‘column_index_number’] 

# Manipulating the index

Manipulating the index
Label-based selection derives its power from the labels in the index. Critically, the index we use is not immutable. We can manipulate the index in any way we see fit.

The set_index() method can be used to do the job. Here is what happens when we set_index to the title field:

In [17]:
df.set_index("Start Date")

First Name  Gender Last Login Time  Salary  Bonus %  \
Start Date                                                       
8/6/1993      Douglas    Male        12:42 PM   97308    6.945   
3/31/1996      Thomas    Male         6:53 AM   61933    4.170   
4/23/1993       Maria  Female        11:17 AM  130590   11.858   
3/4/2005        Jerry    Male         1:00 PM  138705    9.340   
1/24/1998       Larry    Male         4:47 PM  101004    1.389   
...               ...     ...             ...     ...      ...   
11/23/2014      Henry     NaN         6:09 AM  132483   16.655   
1/31/1984     Phillip    Male         6:30 AM   42392   19.675   
5/20/2013     Russell    Male        12:39 PM   96914    1.421   
4/20/2013       Larry    Male         4:45 PM   60500   11.985   
5/15/2012      Albert    Male         6:24 PM  129949   10.169   

           Senior Management                  Team  
Start Date                                          
8/6/1993                True             Marketing  
3/31/1996               True                   NaN  
4/23/1993              False               Finance  
3/4/2005                True               Finance  
1/24/1998               True       Client Services  
...                      ...                   ...  
11/23/2014             False          Distribution  
1/31/1984              False               Finance  
5/20/2013              False               Product  
4/20/2013              False  Business Development  
5/15/2012               True                 Sales  

[1000 rows x 7 columns]

# Summary Functions and Maps

In [25]:
import pandas as pd
import matplotlib.pyplot as plt
df = pd.read_csv("employees.csv")
df

First Name  Gender  Start Date Last Login Time  Salary  Bonus %  \
0      Douglas    Male    8/6/1993        12:42 PM   97308    6.945   
1       Thomas    Male   3/31/1996         6:53 AM   61933    4.170   
2        Maria  Female   4/23/1993        11:17 AM  130590   11.858   
3        Jerry    Male    3/4/2005         1:00 PM  138705    9.340   
4        Larry    Male   1/24/1998         4:47 PM  101004    1.389   
..         ...     ...         ...             ...     ...      ...   
995      Henry     NaN  11/23/2014         6:09 AM  132483   16.655   
996    Phillip    Male   1/31/1984         6:30 AM   42392   19.675   
997    Russell    Male   5/20/2013        12:39 PM   96914    1.421   
998      Larry    Male   4/20/2013         4:45 PM   60500   11.985   
999     Albert    Male   5/15/2012         6:24 PM  129949   10.169   

    Senior Management                  Team  
0                True             Marketing  
1                True                   NaN  
2               False               Finance  
3                True               Finance  
4                True       Client Services  
..                ...                   ...  
995             False          Distribution  
996             False               Finance  
997             False               Product  
998             False  Business Development  
999              True                 Sales  

[1000 rows x 8 columns]

Pandas provides many simple "summary functions" (not an official name) which restructure the data in some useful way. For example, consider the describe() method:

In [26]:
df.Salary.describe()

count      1000.000000
mean      90662.181000
std       32923.693342
min       35013.000000
25%       62613.000000
50%       90428.000000
75%      118740.250000
max      149908.000000
Name: Salary, dtype: float64

This method generates a high-level summary of the attributes of the given column. It is type-aware, meaning that its output changes based on the data type of the input. The output above only makes sense for numerical data; for string data here's what we get:

In [27]:
df.Team.describe()

count                 957
unique                 10
top       Client Services
freq                  106
Name: Team, dtype: object

if you want to get some particular simple summary statistic about a column in a DataFrame or a Series, there is usually a helpful pandas function that makes it happen.

For example, to see the mean of the points allotted (e.g. how well an averagely rated wine does), we can use the mean() function:

In [37]:
df.Salary.mean()

90662.181

In [39]:
df.Team.unique()

array(['Marketing', nan, 'Finance', 'Client Services', 'Legal', 'Product',
       'Engineering', 'Business Development', 'Human Resources', 'Sales',
       'Distribution'], dtype=object)

In [45]:
ds = df.Team.value_counts()

In [48]:
ds

Team
Client Services         106
Finance                 102
Business Development    101
Marketing                98
Product                  95
Sales                    94
Engineering              92
Human Resources          91
Distribution             90
Legal                    88
Name: count, dtype: int64